File: src/algorithms/strings/palindrome/palindrome-check/__tests__/palindrome-check.test.ts
/** Correctness tests for the palindromeCheck function. */

import { describe, it, expect } from "vitest";
import { palindromeCheck } from "../sources/palindrome-check.ts?fn";

describe("palindromeCheck", () => {
  it("returns true for a classic odd-length palindrome", () => {
    expect(palindromeCheck("racecar")).toBe(true);
  });

  it("returns false for a non-palindrome", () => {
    expect(palindromeCheck("hello")).toBe(false);
  });

  it("returns true for a single character", () => {
    expect(palindromeCheck("a")).toBe(true);
  });

  it("returns true for an empty string", () => {
    expect(palindromeCheck("")).toBe(true);
  });

  it("returns false for a two-character non-palindrome", () => {
    expect(palindromeCheck("ab")).toBe(false);
  });

  it("returns true for an odd-length symmetric string", () => {
    expect(palindromeCheck("aba")).toBe(true);
  });

  it("returns true for an even-length palindrome", () => {
    expect(palindromeCheck("abba")).toBe(true);
  });

  it("returns false when only the first and last chars differ", () => {
    expect(palindromeCheck("abca")).toBe(false);
  });

  it("returns true for a string of repeated identical characters", () => {
    expect(palindromeCheck("aaaa")).toBe(true);
  });
});

File: src/algorithms/strings/palindrome/palindrome-check/__tests__/palindrome-check_test.go
package main

import "testing"

func TestPalindromeCheckRacecar(t *testing.T) {
	if !palindromeCheck("racecar") {
		t.Error("expected true for 'racecar'")
	}
}

func TestPalindromeCheckHelloFalse(t *testing.T) {
	if palindromeCheck("hello") {
		t.Error("expected false for 'hello'")
	}
}

func TestPalindromeCheckSingleChar(t *testing.T) {
	if !palindromeCheck("a") {
		t.Error("expected true for single char")
	}
}

func TestPalindromeCheckEmptyString(t *testing.T) {
	if !palindromeCheck("") {
		t.Error("expected true for empty string")
	}
}

func TestPalindromeCheckTwoCharNonPalindrome(t *testing.T) {
	if palindromeCheck("ab") {
		t.Error("expected false for 'ab'")
	}
}

func TestPalindromeCheckOddLengthPalindrome(t *testing.T) {
	if !palindromeCheck("aba") {
		t.Error("expected true for 'aba'")
	}
}

func TestPalindromeCheckEvenLengthPalindrome(t *testing.T) {
	if !palindromeCheck("abba") {
		t.Error("expected true for 'abba'")
	}
}

func TestPalindromeCheckFirstLastDiffer(t *testing.T) {
	if palindromeCheck("abca") {
		t.Error("expected false for 'abca'")
	}
}

func TestPalindromeCheckAllSameChars(t *testing.T) {
	if !palindromeCheck("aaaa") {
		t.Error("expected true for 'aaaa'")
	}
}
